addons/product/models/product_pricelist.py

```diff
@@ -221,7 +221,15 @@ def _compute_price_rule(self, products_qty_partner, date=False, uom_id=False):
                     price = product.price_compute(rule.base)[product.id]
 
                 if price is not False:
-                    price = rule._compute_price(price, price_uom, product, quantity=qty, partner=partner)
+                    # we must take into account the currency of the pricelist compared to the one of the product
+                    def conv_curr_pricelist(price):
+                        return self.currency_id._convert(price, product.currency_id, self.env.company, date, round=False) if price else price
+                    rule_with_currency = rule.with_context({
+                        'price_surcharge': conv_curr_pricelist(rule.price_surcharge),
+                        'price_min_margin': conv_curr_pricelist(rule.price_min_margin),
+                        'price_max_margin': conv_curr_pricelist(rule.price_max_margin),
+                    })
+                    price = rule_with_currency._compute_price(price, price_uom, product, quantity=qty, partner=partner)
```

Comment:
Can't you simply give the date in the context and manage the conversion in `_compute_price` ?
(Or convert the price given to `_compute_price` before the call, so that the margin/surcharge is applied on a price already specified in the pricelist currency)